lib/Auth/Process/EnsureVoMember.php

```diff
@@ -0,0 +1,170 @@
+            return;
+        }
+
+        $attrValues = $this->adapter->getFacilityAttributesValues(
+            $facility,
+            [$this->voDefsAttr, $this->triggerAttr, $this->loginUrlAttr]
+        );
+
+        $triggerAttrValue = $attrValues[$this->triggerAttr];
+        if ($triggerAttrValue === null || $triggerAttrValue === false) {
+            Logger::debug(
+                'perun:EnsureVoMember: skip execution - attribute ' . self::TRIGGER_ATTR . ' is null or false'
+            );
+            return;
+        }
+
+        if (empty($attrValues[$this->voDefsAttr])) {
+            Logger::debug(
+                'perun:EnsureVoMember: skip execution - attribute ' . self::VO_DEFS_ATTR . ' has null or no value'
+            );
+            return;
+        }
+        $voShortName = $attrValues[$this->voDefsAttr];
+
+        $canAccess = $this->adapter->isUserInVo($user, $voShortName);
+        if ($canAccess) {
+            Logger::debug('perun:EnsureVoMember: user allowed to continue');
+        } else {
+            $this->redirect($request, $attrValues[$this->loginUrlAttr], $voShortName);
+        }
+    }
+
+    private function redirect($request, $loginUrl, $voShortName)
+    {
+        if (!empty($voShortName) &&
+            !empty($this->registrarUrl) &&
+            $this->adapter->hasRegistrationFormByVoShortName($voShortName)
+        ) {
+            $this->redirectDirectly($loginUrl, $voShortName);
+        } else {
+            $this->redirectUnapproved($request);
+        }
+    }
+
+    private function redirectDirectly($loginUrl, $voShortName)
+    {
+        HTTP::redirectTrustedURL(
+            $this->registrarUrl,
+            [
+                'vo' => $voShortName,
+                'targetnew' => $loginUrl,
+                'targetexisting' => $loginUrl
+            ]
+        );
+    }
+
+    private function redirectUnapproved($request)
+    {
+        $id = State::saveState($request, 'perunauthorize:Perunauthorize');
+        $url = Module::getModuleURL('perunauthorize/perunauthorize_403.php');
+        if (isset($request['SPMetadata']['InformationURL']['en'])) {
+            HTTP::redirectTrustedURL(
+                $url,
+                [
+                    'StateId' => $id,
+                    'informationURL' => $request['SPMetadata']['InformationURL']['en'],
+                    'administrationContact' => $request['SPMetadata']['administrationContact'],
+                    'serviceName' => $request['SPMetadata']['name']['en']
+                ]
+            );
+        } else {
+            HTTP::redirectTrustedURL(
+                $url,
+                [
+                    'StateId' => $id,
+                    'administrationContact' => $request['SPMetadata']['administrationContact'],
+                    'serviceName' => $request['SPMetadata']['name']['en']
+                ]
+            );
+        }
+    }
```

Comment:
I don't like this construct. I would prefer if the parameters building would be the first block of code, like:
```
$params = [];
$params['StateId'] = $id;
$params['administrationContact'] = $request['SPMetadata']['administrationContact'];
$params['serviceName'] = $request['SPMetadata']['name']['en'];
if (isset($request['SPMetadata']['InformationURL']['en'])) {
    $params['informationURL'] = $request['SPMetadata']['InformationURL']['en'];
}
HTTP::redirectTrustedURL($url, $params);
```

It's a lot shorter :)